File: clevertap-hms/src/test/java/com/clevertap/android/hms/HmsPushProviderTest.java
package com.clevertap.android.hms;

import com.clevertap.android.sdk.pushnotification.CTPushProviderListener;
import com.clevertap.android.sdk.pushnotification.PushConstants;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = 28,
        application = TestApplication.class
)
public class HmsPushProviderTest extends BaseTestCase {

    private CTPushProviderListener ctPushProviderListener;
    private HmsPushProvider pushProvider;
    private TestHmsSdkHandler sdkHandler;

    @Before
    public void setUp() throws Exception {
        super.setUp();
        pushProvider = new HmsPushProvider();
        ctPushProviderListener = mock(CTPushProviderListener.class);
        sdkHandler = new TestHmsSdkHandler();
        when(ctPushProviderListener.context()).thenReturn(application);
        when(ctPushProviderListener.config()).thenReturn(cleverTapInstanceConfig);
        pushProvider.setCTPushListener(ctPushProviderListener);
        pushProvider.setHmsSdkHandler(sdkHandler);
    }

    @Test
    public void testRequestToken() {
        pushProvider.requestToken();
        verify(ctPushProviderListener).onNewToken(HmsTestConstants.HMS_TOKEN, PushConstants.PushType.HPS);
    }

    @Test
    public void testIsAvailable() {
        sdkHandler.setAvailable(false);
        Assert.assertFalse(pushProvider.isAvailable());
        sdkHandler.setAvailable(true);
        Assert.assertTrue(pushProvider.isAvailable());
    }

    @Test
    public void testIsSupported() {
        sdkHandler.setSupported(true);
        Assert.assertTrue(pushProvider.isSupported());
        sdkHandler.setSupported(false);
        Assert.assertFalse(pushProvider.isSupported());
    }
}

File: clevertap-hms/src/test/java/com/clevertap/android/hms/HmsPushProviderTest.kt
package com.clevertap.android.hms

import com.clevertap.android.hms.HmsConstants.MIN_CT_ANDROID_SDK_VERSION
import com.clevertap.android.hms.HmsConstants.HPS
import com.clevertap.android.sdk.pushnotification.CTPushProviderListener
import com.clevertap.android.shared.test.BaseTestCase
import com.clevertap.android.shared.test.TestApplication
import io.mockk.mockk
import io.mockk.verify
import org.junit.*
import org.junit.runner.*
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [28], application = TestApplication::class)
class HmsPushProviderTest : BaseTestCase() {

    private lateinit var ctPushProviderListener: CTPushProviderListener
    private lateinit var pushProvider: HmsPushProvider
    private lateinit var sdkHandler: IHmsSdkHandler

    @Before
    @Throws(Exception::class)
    override fun setUp() {
        super.setUp()
        ctPushProviderListener = mockk<CTPushProviderListener>(relaxed = true)
        pushProvider = HmsPushProvider(ctPushProviderListener, application, cleverTapInstanceConfig)
        sdkHandler = mockk<IHmsSdkHandler>(relaxed = true)
        pushProvider.setHmsSdkHandler(sdkHandler)
    }

    @Test
    fun testRequestToken() {
        pushProvider.requestToken()
        verify(exactly = 1) { sdkHandler.onNewToken() }
        verify(exactly = 1) { ctPushProviderListener.onNewToken(or(isNull(), any()), eq(HPS)) }
    }

    @Test
    fun testIsAvailable() {
        pushProvider.isAvailable
        verify(exactly = 1) { sdkHandler.isAvailable }
    }

    @Test
    fun testIsSupported() {
        pushProvider.isSupported
        verify(exactly = 1) { sdkHandler.isSupported }
    }

    @Test
    fun testGetPushType() {
        Assert.assertEquals(pushProvider.pushType, HPS)
    }

    @Test
    fun minSDKSupportVersionCode() {
        Assert.assertEquals(
            pushProvider.minSDKSupportVersionCode().toLong(),
            MIN_CT_ANDROID_SDK_VERSION.toLong()
        )
    }
}